### ISSUE TITLE
Exports not longer sent via email but accessible in new page  

### ISSUE THREAD

**Issue by janagombitova** [akvo/akvo-flow#2505]
# Context 

When we did our data security audit under the work for becoming compliant with GDPR we realised that the way we make downloading captured data possible is not the best practice.  https://github.com/akvo/akvo-product-design/issues/283 We currently send data exports to users via email. As it can take quite some time (from a few minutes to up to an hour) to generate the data export file depending of the size of the data set, in the past, we decided not to have the user wait in the UI for the file to be generated, but let her freely navigate and then send her the ready file to her email. 

As we want to support our partners in becoming more responsible when handling data, we feel we should change how we make data accessible. We do not want to send data to emails but make sure you can only access it via the UI, respecting your rights and permissions. Once our users download the data to their computers it is up to them how they share it further and handle the data. 

## Current situation
Salim goes to his Flow instance. Decides he wants to work with the dataset of a particular form. He goes to the Reports tab, Export reports page and selects his form. Then he selects which export option suits him the best and hits download. He gets a message telling him that the file is being generated and that he will receive it in his email inbox. A while later (depending on the file size) he receives the email with a link. Once he clicks on the link the data file is downloaded to his computer. 

# Opportunity 

**Note:** This issue holds the entire story, if needed please split the implementation to smaller issues.

**Goal:** How might we make it possible for Salim to download a dataset, not have to wait for it to generate, and once it is ready know that it is ready and download the data from Flow's UI respecting his roles and permissions. 

We decided to create a place in Flow where Salim can get the data from. This place will be close to the current reports page as the actions are linked. Here he will see all the exports he previously created and the newest one on the top. This way he might not need to regenerate an export over and over again if he can just re-download it. He will only see the export he has created and not the exports created by his colleagues on the instance, even though he might have access to the form. We will keep exports that are up to 1 year old, once older we will remove the entry from the UI and the file too (as Salim can always re-download the data set again).

In the UI we will show:  
- Salim goes to the Reports tab and lands on this new page that will hold the generated exports and a button that brings him to the current Export reports page where he selects a form and export option to download
- 1st visit he will not have any generated exports in there, so he will see an empty screen with the prompt to export using the button above 
- To create a new export he hits the button, what brings him to the current exports page. Once he selects his form and export option and hits download, Flow brings him back to the initial page 
- Here he sees
   - the icon of the export type he selected (same as in the current Export reports page)   
   - the name of the exported file: (current naming convention) Export_type-[form ID].[filetype]
   - the full path folder > subfolder > survey > form - this is a dynamic property, so if Salim later 
changes the name or location it needs to reflect the changes 
   - generating icon 
   - something telling him that he will get an email notification... ? 
- As creating can take a while, Salim can navigate away from this page, the generating process will not stop
- If he stays on the page, we need to make sure the generating icon updates properly (without the need of Salim to reload the page)
- Once the file generation is completed, instead of the icon we show the date and time
- With the completion we also send Salim an email to notify him that his data export is ready to download from the UI
- now Salim can click on the file name to download the data. We have decided not to go for a download button to keep the UI clean and the area that is clickable should be distinct enough for Salim to know he can click on it to download
- every time Salim exports a file, the new export comes on top of the list
- We will keep exports that are up to 1 year old, once older we will remove the entry from the UI and the file too (as Salim can always re-download the data set again).
- Salim sees the oldest entries by... (navigation in the list still needs to be defined)

# Next steps 
* [x] finalise the designs 
* [x] test with users
* [ ] change UI according to mock ups 
* [ ] do not save exported file longed than a year/once a file is generated more than a year ago remove it (remove the link and entry from the UI and the file from storage) 
* [ ] ensure folder > survey > form path reflects changes (is a dynamic property)  
* [ ] make it possible to download the export from the UI 
* [ ] update the text to the emails
* [ ] handling data cleaning

# To test with users 
- Does this new UI make sense? Is it understandable? Do you understand how to get a data export and what you see in the list?
- Is the email notification informative? 
- Will you regenerate an export or re-use an export you have done before?
- How long do we need to keep the exported files? We not think 1 year is ok, is it? 

# For later iterations 
- Handling situations when exports fail silently https://github.com/akvo/akvo-flow/issues/1780 --> we should inform the user every time the export fails, but this will be handled in a separate issue and continuously as we find times when exports fail 
- We thought it is useful to show how long it takes to generate an export - this is a nice to have and will also be considered later in the future, not now 
- We will revise the naming convention of the export files as we currently use internal IDs that we do not want to show in the UI, or as least possible

# Related issues 
parent #2544, parent #2610, parent #2611

**Comment by janagombitova** [comment]
# Mock ups 
https://projects.invisionapp.com/share/2VG90WNJS9G#/screens/281499307

# Results from testing 
We tested the designs with three different users from different regions. For the detailed notes you can check here: https://docs.google.com/spreadsheets/d/1ol02GxX7vzUHQoQDo-vBanbuxwI8hELnDnqrMtVhiwM/edit#gid=0

## Insights 
- navigation is clear and understandable as well as the text used 
- we need to make sure in the UI it is clear you can download the file by clicking on the row 
- for some users having also an email notification was good, as while waiting for the file to generate they will go and do something else (most likely outside of Flow) and some found it unneeded. For the first months of the implementation we will keep sending the emails with a generic text informing the file is ready to download from the exports tab. 

# Changes
In this issue we will handle the larger UI changes. The other linked changes are listed below with their respective issues:
- do not save exported file longed than a year/once a file is generated more than a year ago remove it (remove the link and entry from the UI and the file from storage)
- ensure folder > survey > form path reflects changes (is a dynamic property)
- make it possible to download the export from the UI and stop sending the urls via email 
- update the text to the emails #1334
- handling failure of generating an export - https://github.com/akvo/akvo-flow/issues/2543

# Changes for this issue - UI - New Generated exports screen 
This is the initial screen where the user lands after clicking in the main menu on current Reports. The screen holds all the generated exports the user generated. From this page the user can also download the generated exports. 

**Expected behaviour:** https://projects.invisionapp.com/share/2VG90WNJS9G#/screens/281499307

* [ ] New Generated exports page, if there are no exports, Salim sees the blank page, with text in the middle and the New export button 
* [ ] New export button brings him to the current Export reports page where he selects the surrvey > form, export type --> no changes to current implementation 
* [ ] change in Export reports page the `Download` button to `Generate` 
* [ ] Once Salim generates a file, bring him directly back tot he Generated exports page 
* [ ] the newly generating export file is on the top of the list - with its properties and a spinning wheel and `Generating`. The row is not clickable. 
   - Icon that is the same and based on the export type the user decided to generate   
   - File name - as we have currently: DATA_CLEANING-[form ID] 
   - root folder > subfolder > survey > form where the file is coming from 
* [ ] if Salim navigates away while the file generation is in progress, the generation of the file should not stop
* [ ] if Salim remains waiting on the page for the file generation to complete, he should see the change immediately (no need to refresh the page) 
* [ ] Once file is generated show the date and the row becomes clickable. 
* [ ] By clicking on the full row the user can download the file 
* [ ] Hovering over the row should indicate that you can download 
* [ ] Once a new export is generated indicate it differently from the list, so the user can keep track of which exports he/she already downloaded 

<img width="944" alt="screen shot 2018-03-16 at 16 36 34" src="https://user-images.githubusercontent.com/12456965/37530260-73c27080-2939-11e8-86c4-7b308a1c5348.png">

<img width="962" alt="screen shot 2018-03-16 at 17 00 28" src="https://user-images.githubusercontent.com/12456965/37531121-87ca4d8a-293b-11e8-9711-2f1a80a079af.png">

**Comment by loicsans** [comment]
I'm on this one for now, building the new UI elements necessary

**Comment by stellanl** [comment]
I suggest we create a new REST endpoint: /exports (or /reports ?) that will be used by
 1. By the UI to find out which reports have been created, their state and where to download the result from. (optionally?) filtered by user.
 2. By the report engine to create and update the state of the export.
 3. By a background cleanup job that removes the exports when they get too old.

**Comment by janagombitova** [comment]
@loicsans, @stellanl and @muloem I just realised that there is one thing missing we did not think of for the initial implementation: How will we distinguish in the UI that a data export might have been filtered out based on submission date? 

Salim can generate a data export that he filters out on the submissions date. A few days ago, he might have requested the data export for the same form but with a different date filter set. How will he know the different when looking at his list? 

What do you think? And is this something we can add to the initial implementation or should it be a separate issue for the follow up round?

**Comment by stellanl** [comment]
I'll add the start and end dates to the data structures right away. We'll see how much will make it into the UI for the first release.

**Comment by kymni** [comment]
@muloem, yes that should be the case. Investigating

**Comment by janagombitova** [comment]
Hey everyone, this is coming together very nicely. I could not resist and had to check it out and noticed a few inconsistencies with the implementation and the final designs that should be addressed: 

![screen shot 2018-07-04 at 08 33 38](https://user-images.githubusercontent.com/12456965/42260502-fb7d531a-7f64-11e8-9c49-46d98d8b3e51.png)

# Download button
The 'Download' button should now be 'Generate' as here is where Salim asks for the exports, but in the list he actually downloads them. Saying download on this button can confuse Salim on what the role of the list is. 

# Navigating back to the list
Once Salim selects which form he wants to export and hits the current 'Download' button, he gets a message about the email notification. This should be removed, but we can keep it for now. However, once he clicks 'ok' on the message he stays on the New export page where he can generate exports. He should be brought to the list. Having to manually go back to the list makes him do extra work, plus it makes it even more clear that his export is going to be ready for download for the list. 

# Alignment in the list
![screen shot 2018-07-04 at 08 37 35](https://user-images.githubusercontent.com/12456965/42260638-864091c4-7f65-11e8-85ef-1567d831686d.png)

Behind the Generating/Date is a large gap. I was expecting some button or something to show there, but nothing came and I kept wondering a realised it is just empty. Can we change the alignment here somehow to get rid of the gap @loicsans ?

**Comment by janagombitova** [comment]
Nice. Should we create a separate issue for the **UI Alignment** for the next release so we can then ceremonially close this one? :)

**Comment by loicsans** [comment]
Agreed with @kymni, let's fix this for the next release then
